File: c/src/proto/lineage_replace_faults.c
#include "auth/lineage_replace_faults.h"

lineage_replace_recovery_facts_t lineage_replace_facts_at_write_cut(lineage_replace_write_cut_t cut) {
    lineage_replace_recovery_facts_t facts = {false, false, false, false, false, false};
    facts.record_integrity_valid = true;

    switch (cut) {
        case LINEAGE_REPLACE_CUT_BEFORE_BEGIN:
            facts.predecessor_active = true;
            break;
        case LINEAGE_REPLACE_CUT_AFTER_PENDING_MARKER:
            facts.predecessor_active = true;
            facts.replacement_pending = true;
            break;
        case LINEAGE_REPLACE_CUT_AFTER_SUCCESSOR_ACTIVATION:
            facts.predecessor_active = true;
            facts.replacement_pending = true;
            facts.successor_active = true;
            break;
        case LINEAGE_REPLACE_CUT_AFTER_PREDECESSOR_RETIREMENT:
            facts.replacement_pending = true;
            facts.successor_active = true;
            facts.predecessor_retired = true;
            break;
        case LINEAGE_REPLACE_CUT_AFTER_INVALIDATIONS:
            facts.replacement_pending = true;
            facts.successor_active = true;
            facts.predecessor_retired = true;
            facts.invalidations_complete = true;
            break;
        case LINEAGE_REPLACE_CUT_AFTER_COMMIT_MARKER_CLEAR:
            facts.successor_active = true;
            facts.predecessor_retired = true;
            facts.invalidations_complete = true;
            break;
        default:
            facts.record_integrity_valid = false;
            break;
    }

    return facts;
}

File: rust/crates/proto/src/lineage_replace_faults.rs
//! Deterministic, storage-neutral write-cut model for `LINEAGE_REPLACE`.
//!
//! Each cut point produces the normalized restart facts that a future storage
//! adapter would expose if execution stopped at that logical boundary. This is
//! a model of ordering requirements, not evidence about any filesystem, flash
//! device, journal, secure element, or physical power-loss behavior.

use crate::lineage_replace_recovery::LineageReplaceRecoveryFacts;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum LineageReplaceWriteCut {
    BeforeBegin,
    AfterPendingMarker,
    AfterSuccessorActivation,
    AfterPredecessorRetirement,
    AfterInvalidations,
    AfterCommitMarkerClear,
}

pub fn facts_at_write_cut(cut: LineageReplaceWriteCut) -> LineageReplaceRecoveryFacts {
    match cut {
        LineageReplaceWriteCut::BeforeBegin => LineageReplaceRecoveryFacts {
            record_integrity_valid: true,
            predecessor_active: true,
            replacement_pending: false,
            successor_active: false,
            predecessor_retired: false,
            invalidations_complete: false,
        },
        LineageReplaceWriteCut::AfterPendingMarker => LineageReplaceRecoveryFacts {
            record_integrity_valid: true,
            predecessor_active: true,
            replacement_pending: true,
            successor_active: false,
            predecessor_retired: false,
            invalidations_complete: false,
        },
        LineageReplaceWriteCut::AfterSuccessorActivation => LineageReplaceRecoveryFacts {
            record_integrity_valid: true,
            predecessor_active: true,
            replacement_pending: true,
            successor_active: true,
            predecessor_retired: false,
            invalidations_complete: false,
        },
        LineageReplaceWriteCut::AfterPredecessorRetirement => LineageReplaceRecoveryFacts {
            record_integrity_valid: true,
            predecessor_active: false,
            replacement_pending: true,
            successor_active: true,
            predecessor_retired: true,
            invalidations_complete: false,
        },
        LineageReplaceWriteCut::AfterInvalidations => LineageReplaceRecoveryFacts {
            record_integrity_valid: true,
            predecessor_active: false,
            replacement_pending: true,
            successor_active: true,
            predecessor_retired: true,
            invalidations_complete: true,
        },
        LineageReplaceWriteCut::AfterCommitMarkerClear => LineageReplaceRecoveryFacts {
            record_integrity_valid: true,
            predecessor_active: false,
            replacement_pending: false,
            successor_active: true,
            predecessor_retired: true,
            invalidations_complete: true,
        },
    }
}
